subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/DomainSize.java

```diff
@@ -0,0 +1,98 @@
+/*
+ *  Copyright 2023 Budapest University of Technology and Economics
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package hu.bme.mit.theta.core.type;
+
+import com.google.common.base.Objects;
+
+import java.math.BigInteger;
+import java.util.function.BinaryOperator;
+
+public class DomainSize {
+
+	private final BigInteger finiteSize;
+
+	private DomainSize(BigInteger value) {
+		finiteSize = value;
+	}
+
+	public static DomainSize of(BigInteger val) {
+		return new DomainSize(val);
+	}
+
+	public static DomainSize of(long val) {
+		return of(BigInteger.valueOf(val));
+	}
+
+	public static final DomainSize INFINITY = of(-1);
+	public static final DomainSize ZERO = of(0);
+	public static final DomainSize ONE = of(1);
+	public static final DomainSize TWO = of(2);
+
+	private static DomainSize infiniteForAnyInfiniteApplyElse(DomainSize left, DomainSize right, BinaryOperator<BigInteger> operator) {
+		if (left.isInfinite() || right.isInfinite())
+			return INFINITY;
+		return of(operator.apply(left.finiteSize, right.finiteSize));
+	}
+
+	public boolean isInfinite() {
+		return equals(INFINITY);
+	}
+
+	public boolean isBiggerThan(long limit) {
+		return finiteSize.compareTo(BigInteger.valueOf(limit)) > 0;
+	}
```

Comment:
I think it will not work for `INFINITY` as intended.

Reply:
Done in a315e11